source/working-with-bson.txt

```diff
@@ -0,0 +1,399 @@
+   statements, you must capture each intermediate result. Additionally, the last
+   statement within the loop body should assign its result back to the in_array
+   object, so that the loop restarts in a consistent state:
+
+   .. code-block:: cpp
+
+      for (auto && e : {1, 2, 3}) {
+         auto open_state = in_array << builder::stream::open_document;
+         auto temp_state = open_state << "key" << e;
+         in_array = temp_state << builder::stream::close_document;
+      }
+
+.. _cpp-bson-value:
+
+Owning BSON Documents
+~~~~~~~~~~~~~~~~~~~~~
+
+`bsoncxx::document::value <https://github.com/mongodb/mongo-cxx-driver/blob/master/src/bsoncxx/include/bsoncxx/v_noabi/bsoncxx/document/value.hpp>`__
+
+This type represents an actual BSON document, one that owns its buffer of
+data.  These documents can be constructed from a builder by calling
+``extract()``:
+
+.. code-block:: cpp
+
+   bsoncxx::document::value basic_doc{basic_builder.extract()};
+   bsoncxx::document::value stream_doc{stream_builder.extract()};
+
+After calling ``extract()`` the builder is in a moved-from state,
+and should not be used.
+
+It's possible to create a ``bsoncxx::document::value`` in a single line using
+the stream builder interface and the ``finalize`` token. ``finalize`` returns a
+``document::value`` from a temporary stream builder:
+
+.. code-block:: cpp
+
+   // { "finalize" : "is nifty" }
+   bsoncxx::document::value one_line = bsoncxx::builder::stream::document{} << "finalize" << "is nifty" << bsoncxx::builder::stream::finalize;
+
+.. _cpp-bson-view:
+
+Non-owning BSON Documents (views)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+`bsoncxx::document::view <https://github.com/mongodb/mongo-cxx-driver/blob/master/src/bsoncxx/include/bsoncxx/v_noabi/bsoncxx/document/view.hpp>`__
+
+This type is a view into an owning ``bsoncxx::document::value``.
+
+.. code-block:: cpp
+
+   bsoncxx::document::view document_view{document_value.view()};
+
+A ``document::value`` also implicitly converts to a ``document::view``:
+
+.. code-block:: cpp
+
+   bsoncxx::document::view document_view{document_value};
+
+In performance-critical code, passing views around is preferable to using
+values because we can avoid excess copying. Also, passing a view of a
+document allows us to use the document multiple times:
+
+.. code-block:: cpp
+
+   // { "copies" : { "$gt" : 100 } }
+   auto query_value = document{} << "copies" << open_document << "$gt" << 100 << close_document << finalize;
+
+   // Run the same query across different collections
+   auto collection1 = db["science_fiction"];
+   auto cursor1 = collection1.find(query_value.view());
+
+   auto collection2 = db["cookbooks"];
+   auto cursor2 = collection2.find(query_value.view());
+
+.. _cpp-bson-view_or_value:
+
+Optionally-owning BSON Documents (view_or_value)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Many driver methods take a document::view_or_value parameter, for example,
```

Comment:
S: small change (that's maybe outside the scope of this review), but I think all mentions of the `document::` types should be monospaced
```suggestion
Many driver methods take a ``document::view_or_value`` parameter, for example,
```